File: cpp/src/0202.cpp
/*
[202. Happy Number (Easy)](https://leetcode.com/problems/happy-number/)

@topic: Hash Table
@topic: Math
@topic: Two Pointers

Constraints:
  - `1 <= n <= 2^31 - 1`
*/

class Solution {
   public:
    bool isHappy(int n) {
        while (true) {
            switch (n) {
                case 1:
                    return true;
                case 4:
                case 16:
                case 20:
                case 37:
                case 42:
                case 58:
                case 89:
                case 145:
                    return false;
                default:
                    n = next(n);
                    break;
            }
        }
    }

    // bool isHappy(int n) {
    //     int slow = n;
    //     int fast = n;
    //     while (true) {
    //         slow = next(slow);
    //         fast = next(fast);
    //         fast = next(fast);
    //         if (fast == 1) {
    //             return true;
    //         }
    //         if (slow == fast) {
    //             return slow == 1;
    //         }
    //     }
    // }

   private:
    int next(int n) {
        int s = 0;
        while (n) {
            int m = n % 10;
            n /= 10;
            s += m * m;
        }
        return s;
    }
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    Solution o;

    vector<tuple<int, bool>> CASES = {
        {19, true},
        {2, false},
    };

    for (auto& [n, excepted] : CASES) {
        assert(o.isHappy(n) == excepted);
    }

    auto start = system_clock::now();
    for (auto& [n, _] : CASES) {
        for (int i = 0; i < 1000000; ++i) {
            o.isHappy(n);
        }
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}

File: python/src/0202.py
class Solution:
    # def isHappy(self, n: int) -> bool:
    #     fast = slow = n
    #     while True:
    #         slow = self.next(slow)
    #         fast = self.next(fast)
    #         fast = self.next(fast)
    #         if fast == 1:
    #             return True
    #         if slow == fast:
    #             return slow == 1

    def isHappy(self, n: int) -> bool:
        while True:
            if n == 1:
                return True
            if n in {4, 16, 20, 37, 42, 58, 89, 145}:
                return False
            n = self.next(n)

    def next(self, n):
        s = 0
        while n:
            n, m = divmod(n, 10)
            s += m * m
        return s


if __name__ == '__main__':
    import unittest

    CASES = (
        (1, True),
        (2, False),
        (3, False),
        (4, False),
        (5, False),
        (6, False),
        (7, True),
        (8, False),
        (9, False),
        (10, True),
        (11, False),
        (19, True),
        (116, False),
    )

    class Test(unittest.TestCase):
        def test(self):
            isHappy = Solution().isHappy
            for n, excepted in CASES:
                actual = isHappy(n)
                self.assertEqual(actual, excepted)

    unittest.main()

    from timeit import timeit
    isHappy = Solution().isHappy
    for n, excepted in CASES:
        print(timeit(lambda: isHappy(n), number=100000))
